doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -659,168 +734,476 @@ insert in a concise way an Ltac2 variable of type :n:`constr` into a Coq term.
+   Every possible match for the hypotheses is evaluated until the right-hand
+   side succeeds.  Note that `H1` and `H2` are never matched to the same hypothesis.
+   Observe that the number of permutations can grow as the factorial
+   of the number of hypotheses and hypothesis patterns.
 
-- :n:`terminal(s = @string)`:
+   .. coqtop:: all abort
 
-  + parses the string :n:`s` as a keyword, if it is already a
-    keyword, otherwise as an :n:`@ident`. Returns `()`.
+      Goal forall A B : Prop, A -> B -> (A->B).
+      intros A B H.
+      match! goal with
+      | [ h1 : _, h2 : _ |- _ ] =>
+         print (concat (of_string "match ")
+               (concat (of_constr (Control.hyp h1))
+               (concat (of_string " ")
+               (of_constr (Control.hyp h2)))));
+         fail
+      | [ |- _ ] => ()
+      end.
 
-- :n:`seq(@ltac2_scope__1, ..., @ltac2_scope__2)`:
 
-  + parses :n:`scope__1`, ..., :n:`scope__n` in this order, and produces a tuple made
-    out of the parsed values in the same order. As an optimization, all
-    subscopes of the form :n:`STRING` are left out of the returned tuple, instead
-    of returning a useless unit value. It is forbidden for the various
-    subscopes to refer to the global entry using :n:`self` or :n:`next`.
+Match on values
+~~~~~~~~~~~~~~~
 
-A few other specific scopes exist to handle Ltac1-like syntax, but their use is
-discouraged and they are thus not documented.
+.. tacn:: match @ltac2_expr5 with {? @ltac2_branches } end
+   :name: match (Ltac2)
 
-For now there is no way to declare new scopes from Ltac2 side, but this is
-planned.
+   Matches a value, akin to the OCaml `match` construct.  By itself, it doesn't cause backtracking
+   as do the `*match*!` and `*match*! goal` constructs.
 
-Notations
-~~~~~~~~~
+   .. insertprodn ltac2_branches atomic_tac2pat
 
-The Ltac2 parser can be extended with syntactic notations.
+   .. prodn::
+      ltac2_branches ::= {? %| } {+| @tac2pat1 => @ltac2_expr }
+      tac2pat1 ::= @qualid {+ @tac2pat0 }
+      | @qualid
+      | [ ]
+      | @tac2pat0 :: @tac2pat0
+      | @tac2pat0
+      tac2pat0 ::= _
+      | ()
+      | @qualid
+      | ( {? @atomic_tac2pat } )
+      atomic_tac2pat ::= @tac2pat1 : @ltac2_type
+      | @tac2pat1 , {*, @tac2pat1 }
+      | @tac2pat1
 
-.. cmd:: Ltac2 Notation {+ {| @lident (@ltac2_scope) | @string } } {? : @int} := @ltac2_term
+.. note::
+
+   For now, deep pattern matching is not implemented.
+
+
+.. _ltac2_notations:
+
+Notations
+---------
+
+.. cmd:: Ltac2 Notation @string {* @name ( @ltac2_scope ) } {? : @int } := @ltac2_expr
    :name: Ltac2 Notation
 
-   A Ltac2 notation adds a parsing rule to the Ltac2 grammar, which is expanded
+   .. todo seems like name maybe should use lident rather than ident, considering:
+
+      Ltac2 Notation "ex1" X(constr) := print (of_constr X).
+      ex1 1.
+
+      Unbound constructor X
+
+      This works fine with lower-case "x" in place of "X"
```

Comment:
Indeed, maybe submit an issue for this?

Reply:
The user-friendly thing would be to give a user-friendly error message like "Ltac2 identifiers must begin with a lower-case letter".  https://github.com/coq/coq/issues/12804